src/Compilers/CSharp/Portable/Symbols/ModuleSymbol.cs

```diff
@@ -345,11 +345,45 @@ public NamespaceSymbol GetModuleNamespace(INamespaceSymbol namespaceSymbol)
                 throw new ArgumentNullException(nameof(namespaceSymbol));
             }
 
-            var moduleNs = namespaceSymbol as NamespaceSymbol;
-            if ((object)moduleNs != null && moduleNs.Extent.Kind == NamespaceKind.Module && moduleNs.ContainingModule == this)
+            if (namespaceSymbol.NamespaceKind == NamespaceKind.Module)
+            {
+                var moduleNs = (namespaceSymbol as PublicModel.NamespaceSymbol)?.UnderlyingNamespaceSymbol;
+                if ((object)moduleNs != null && moduleNs.ContainingModule == this)
+                {
+                    // this is already the correct module namespace
+                    return moduleNs;
+                }
+            }
+
+            if (namespaceSymbol.IsGlobalNamespace || (object)namespaceSymbol.ContainingNamespace == null)
+            {
+                return this.GlobalNamespace;
+            }
+            else
+            {
+                var cns = GetModuleNamespace(namespaceSymbol.ContainingNamespace);
+                if ((object)cns != null)
```

Comment:
nit: consider `return cns?.GetNestedNamespace(namespaceSymbol.Name);` (would also be applicable below)